zstd/seqdec_amd64.go

```diff
@@ -101,3 +101,104 @@ func (s *sequenceDecs) decode(seqs []seqVals) error {
 	}
 	return err
 }
+
+type executeAsmContext struct {
+	seqs        []seqVals
+	seqIndex    int
+	out         []byte
+	literals    []byte
+	outPosition int
+	litPosition int
+}
+
+// sequenceDecs_executeSimple_amd64 implements the main loop of sequenceDecs.executeSimple in x86 asm.
+//
+// Please refer to seqdec_generic.go for the reference implementation.
+//go:noescape
+func sequenceDecs_executeSimple_amd64(ctx *executeAsmContext) bool
+
+// executeSimple handles cases when no history nor dictionary are used.
+func (s *sequenceDecs) executeSimple(seqs []seqVals) error {
+	// Ensure we have enough output size...
+	if len(s.out)+s.seqSize > cap(s.out) {
+		addBytes := s.seqSize + len(s.out)
+		s.out = append(s.out, make([]byte, addBytes)...)
+		s.out = s.out[:len(s.out)-addBytes]
+	}
+
+	if debugDecoder {
+		printf("Execute %d seqs with literals: %d into %d bytes\n", len(seqs), len(s.literals), s.seqSize)
+	}
+
+	var t = len(s.out)
+	out := s.out[:t+s.seqSize]
+
+	ctx := executeAsmContext{
+		seqs:     seqs,
+		seqIndex: 0,
+		out:      out,
+	}
+
+	for ctx.seqIndex < len(seqs) {
+		ctx.outPosition = t
+		ctx.litPosition = 0
+		ctx.literals = s.literals
+		completed := sequenceDecs_executeSimple_amd64(&ctx)
```

Comment:
So are there any cases where this returns and it is not an error?

As far as I understand this loop to just be eliminated.

Reply:
True, removed. Just had to add two extra checks and tests pass.